shared/quantum/codeql/quantum/experimental/Model.qll

```diff
@@ -0,0 +1,2596 @@
+    TSymmetricKeyType() or
+    TAsymmetricKeyType() or
+    TUnknownKeyType()
+
+  class KeyArtifactType extends TKeyArtifactType {
+    string toString() {
+      this = TSymmetricKeyType() and result = "Symmetric"
+      or
+      this = TAsymmetricKeyType() and result = "Asymmetric"
+      or
+      this = TUnknownKeyType() and result = "Unknown"
+    }
+  }
+
+  abstract private class KeyArtifactInstance extends ArtifactInstance {
+    abstract KeyArtifactType getKeyType();
+  }
+
+  final class KeyArtifactOutputInstance extends KeyArtifactInstance, OutputArtifactInstance {
+    KeyCreationOperationInstance creator;
+
+    KeyArtifactOutputInstance() { Input::dfn_to_element(creator.getOutputKeyArtifact()) = this }
+
+    final KeyCreationOperationInstance getCreator() { result = creator }
+
+    override KeyArtifactType getKeyType() { result = creator.getOutputKeyType() }
+
+    override DataFlowNode getOutputNode() { result = creator.getOutputKeyArtifact() }
+  }
+
+  final class KeyArtifactConsumer extends KeyArtifactInstance, ArtifactConsumerAndInstance {
+    ConsumerInputDataFlowNode inputNode;
+
+    // TODO: key type hint? e.g. hint: private || public
+    KeyArtifactConsumer() {
+      (
+        exists(KeyOperationInstance op | inputNode = op.getKeyConsumer())
+        or
+        exists(MACOperationInstance op | inputNode = op.getKeyConsumer())
+        or
+        exists(KeyAgreementSecretGenerationOperationInstance op |
+          inputNode = op.getServerKeyConsumer() or
+          inputNode = op.getPeerKeyConsumer()
+        )
+      ) and
+      this = Input::dfn_to_element(inputNode)
+    }
+
+    override KeyArtifactType getKeyType() { result instanceof TUnknownKeyType } // A consumer node does not have a key type, refer to source (TODO: refine, should this be none())
+
+    final override ConsumerInputDataFlowNode getInputNode() { result = inputNode }
+  }
+
+  /**
+   * The `KeyOpAlg` module defines key operation algorithms types (e.g., symmetric ciphers, signatures, etc.)
+   * and provides mapping of those types to string names and structural properties.
+   */
+  module KeyOpAlg {
+    /**
+     * An algorithm used in key operations.
+     */
+    newtype TAlgorithm =
+      TSymmetricCipher(TSymmetricCipherType t) or
+      TAsymmetricCipher(TAsymmetricCipherType t) or
+      TSignature(TSignatureAlgorithmType t) or
+      TKeyEncapsulation(TKEMAlgorithmType t) or
+      TUnknownKeyOperationAlgorithmType()
+
+    // Parameterized algorithm types
+    newtype TSymmetricCipherType =
+      AES() or
+      ARIA() or
+      BLOWFISH() or
+      CAMELLIA() or
+      CAST5() or
+      CHACHA20() or
+      DES() or
+      DESX() or
+      GOST() or
+      IDEA() or
+      KUZNYECHIK() or
+      MAGMA() or
+      TripleDES() or
+      DoubleDES() or
+      RC2() or
+      RC4() or
+      RC5() or
+      SEED() or
+      SM4() or
+      OtherSymmetricCipherType()
+
+    newtype TAsymmetricCipherType =
```

Comment:
Some of these types can be made private.
I think it is best practice to avoid exposing IPA types if possible.

Reply:
You don't have to - these are just generic recommendations 😄 
If you think it is best to keep it as it is - then definitely do so!